src/overlays/actors/ovl_Door_Ana/z_door_ana.c

```diff
@@ -21,20 +24,169 @@ const ActorInit Door_Ana_InitVars = {
+
+    if (grottoType == DOORANA_TYPE_HIDDEN) {
+        Collider_DestroyCylinder(globalCtx, &this->bombCollider);
+    }
+}
+
+void DoorAna_WaitClosed(DoorAna* this, GlobalContext* globalCtx) {
+    s32 grottoIsOpen = 0;
+    u32 grottoType = GET_DOORANA_TYPE(this);
+
+    if (grottoType == DOORANA_TYPE_UNK) {
+        // in OOT decomp its marked as open with storms, but does not seem to open with storms in MM
+        if ((this->actor.xyzDistToPlayerSq < 40000.0f) && (func_800F13E8(globalCtx, 5))) {
+            grottoIsOpen = 1;
+            this->actor.flags &= ~0x10; // always update OFF
+        }
+
+    } else {
+        if (this->bombCollider.base.acFlags & AC_HIT) { // bomb collision
+            grottoIsOpen = 1;
+            Collider_DestroyCylinder(globalCtx, &this->bombCollider);
+
+        } else {
+            Collider_UpdateCylinder(&this->actor, &this->bombCollider);
+            CollisionCheck_SetAC(globalCtx, &globalCtx->colCheckCtx, &this->bombCollider.base);
+        }
+    }
+
+    if (grottoIsOpen) {
+        this->actor.params &= ~DOORANA_TYPE_BITRANGE;
+        DoorAna_SetupAction(this, DoorAna_WaitOpen);
+        play_sound(0x4802); // CORRECT_CHIME sfx
+    }
+
+    func_800B8C50(&this->actor, globalCtx);
+}
+
+void DoorAna_WaitOpen(DoorAna* this, GlobalContext* globalCtx) {
+    ActorPlayer* player = PLAYER;
+    s32 dooranaType = GET_DOORANA_TYPE(this);
+    s8 pad[4];
+    s32 entranceIndex;
+    f32 yDist;
+
+    if (Math_StepToF( &this->actor.scale, 0.01f, 0.001f) != 0) { 
+        if ((this->actor.targetMode != 0) && (globalCtx->unk18875 == 0) && (globalCtx->unk18B4A == 0) &&
+          (player->stateFlags1 & 0x80000000) && (player->unkAE7 == 0)) {
+
+            if (dooranaType == DOORANA_TYPE_ADJACENT) {
+                // 300 uses scene exit addresses, not static DoorAna entrance addresses, 
+                // eg. deku playground gets address in the NCT scene exit table
+
+                entranceIndex = GET_DOORANA_ADJACENT_ENTRANCE(this);
+                globalCtx->nextEntranceIndex = globalCtx->setupExitList[entranceIndex];
+                lblUnk_808E03B8: ; // required to convince compiler to not use delay slot
+
+            } else { 
+                // unused in vanilla, the highest params bits can directly index an address
+                entranceIndex = GET_DOORANA_DIRECT_ENTRANCE(this);
+
+                func_80169E6C(globalCtx, 3, 0x4FF);
+
+                gSaveContext.extra.previousPlayerHeight = this->actor.world.pos.y;
+                gSaveContext.extra.previousPlayerYRot   = this->actor.home.rot.y;
+
+                // save the params lower byte for En_Torch to decide what item to use in the grotto chest
+                gSaveContext.extra.unk87 = GET_DOORANA_ITEMFLAGS(this);
+
+                // most grottos in the game use their zrotation as their entrance index, not params
+                if (DOORANA_TYPE_ROTATION_ENTRANCE(entranceIndex)) {
```

Comment:
I don't think this is a good use of a macro, since all it does it test a local variable. It would be clearer if there was no macro, and the comment read something like "The params can only fit up to index 7. If the index was 0 (it could not fit in there), use the one stored in z rotation."